crates/sui-replay-2/src/data-stores/file_system_store.rs

```diff
@@ -0,0 +1,636 @@
+            protocol_version: epoch_data.protocol_version,
+            rgp: epoch_data.rgp,
+            start_timestamp: epoch_data.start_timestamp,
+        }
+    }
+}
+
+/// File system implementation of the replay interfaces
+pub struct FileSystemStore {
+    node: Node,
+    base_path: PathBuf,
+    metrics: FsStoreMetrics,
+    // In-memory caches of version mappings per object id
+    root_versions_map: RwLock<BTreeMap<ObjectID, BTreeMap<u64, u64>>>,
+    checkpoint_versions_map: RwLock<BTreeMap<ObjectID, BTreeMap<u64, u64>>>,
+}
+
+impl FileSystemStore {
+    pub fn new(node: Node) -> Result<Self, anyhow::Error> {
+        let home_dir = std::env::var("HOME")
+            .or_else(|_| std::env::var("USERPROFILE"))
+            .map_err(|_| anyhow!("Unable to determine home directory"))?;
+        let base_path = PathBuf::from(home_dir).join(".replay_data_store");
+        Ok(Self {
+            node,
+            base_path,
+            metrics: FsStoreMetrics::default(),
+            root_versions_map: RwLock::new(BTreeMap::new()),
+            checkpoint_versions_map: RwLock::new(BTreeMap::new()),
+        })
+    }
+
+    fn node_dir(&self) -> PathBuf {
+        self.base_path.join(self.node.node_dir())
+    }
+
+    fn transaction_dir(&self) -> PathBuf {
+        self.node_dir().join("transaction")
+    }
+
+    fn epoch_dir(&self) -> PathBuf {
+        self.node_dir().join("epoch")
+    }
+
+    fn objects_dir(&self) -> PathBuf {
+        self.node_dir().join("objects")
+    }
+
+    fn root_versions_path(&self, object_id: &ObjectID) -> PathBuf {
+        self.objects_dir()
+            .join(object_id.to_string())
+            .join("root_versions")
+    }
+
+    fn checkpoint_versions_path(&self, object_id: &ObjectID) -> PathBuf {
+        self.objects_dir()
+            .join(object_id.to_string())
+            .join("checkpoint_versions")
+    }
+
+    fn read_bcs_file<T: serde::de::DeserializeOwned>(
+        &self,
+        path: &Path,
+    ) -> Result<T, anyhow::Error> {
+        let bytes =
+            fs::read(path).with_context(|| format!("Failed to read file: {}", path.display()))?;
+        bcs::from_bytes(&bytes)
+            .with_context(|| format!("Failed to deserialize BCS data from: {}", path.display()))
+    }
+
+    fn write_bcs_file<T: serde::Serialize>(
+        &self,
+        path: &Path,
+        data: &T,
+    ) -> Result<(), anyhow::Error> {
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)
+                .with_context(|| format!("Failed to create directory: {}", parent.display()))?;
+        }
+        let bytes = bcs::to_bytes(data)
+            .with_context(|| format!("Failed to serialize BCS data for: {}", path.display()))?;
+        fs::write(path, bytes)
+            .with_context(|| format!("Failed to write file: {}", path.display()))?;
+        Ok(())
+    }
+
+    fn read_version_mapping(&self, path: &Path) -> Result<BTreeMap<u64, u64>, anyhow::Error> {
```

Comment:
Would it be nicer/easier to handle the CSV writing/reading using the [`csv`](https://docs.rs/csv/latest/csv/) crate?